docs/arch/adr-011-kurl-addon-operator.md

```diff
@@ -0,0 +1,96 @@
+
+The operator will be implemented with the following objectives in mind:
+
+- It will be scaffolded using [kubebuilder](https://book.kubebuilder.io/introduction.html).
+- It will be developed using [SOLID](https://en.wikipedia.org/wiki/SOLID) principles and operator best practices.
+- It will distribution agnostic. I.e. it will not be tightly coupled to the kURL Kubernetes distribution.
+- It will manage application add-ons _only_ and _exclude_ managing CRI and CNI resources.
+- It will install and confiure the KOTS admin console.
+- It can be installed on an existing [kURL](https://kurl.sh/) cluster with only Kubernetes, containerd and flannel.
+- The installation mechanism for add-ons will be through the use of Helm Charts.
+- By default the add-on configuration will be opinionated by us but we will enable the end-user to override the configuration for maximum flexibility. Vendors and users will be able to override the configuration using a standard [Helm values](https://helm.sh/docs/chart_template_guide/values_files/) file.
+- It will rely as much as possible on Helm for managing the lifecyle of the add-ons.
+
+### The `Installer` API
+
+The `Installer` API defines a resource for installing specific add-ons.
+
+```yaml
+apiVersion: cluster.kurl.sh/v2beta1
+kind: Installer
+metadata:
+  name: installer-sample
+spec:
+  channel: "1.27"
+  minio:
+    enabled: true
+    configOverride:
+        persistence:
+            enabled: true
+        replicas: 1
+        mode: standalone
+
+```
+
+### The Controllers
+
+The operator will have at minimum one controller which ensures the `Installer` CR is reconciled with the cluster periodically. The `Installer` controller (main controller) will delegate responsiblity for managing add-ons to other controllers. The reconcilation for the main controller will consist of determining the add-ons to be managed from the `Installer` CR and creating a controller for each add-on (addon-on controller).
+
+The add-on controller will be responsible for
+- verifying the requirements for the add-on are met
+- managing the add-on itself
+- ensure add-on is in desired state
+- update the add-on CR status
+
+### Managing the Add-ons
+
+How the add-ons will be managed is still an open question. We have a couple of options:
+
+_Use the Helm Controller provided by Flux_
+
+The [Flux Helm Controller](https://fluxcd.io/flux/components/helm/helmreleases/) affords us the following benefits:
+- Stable [HelmRelease API](https://fluxcd.io/flux/components/helm/api/)
+- Helm release [dependencies](https://fluxcd.io/flux/components/helm/helmreleases/#helmrelease-dependencies)
+- Support for Helm install, upgrade, test, rollback and uninstall
+- Performs Helm uninstall on `HelmRelease` CR removal
+- Chart values override\
+- Pull charts from Helm repository directly
+
+_Develop our own Homebrew Helm Controller_
+
+This approach affords us with a controller that meets our specific needs and avoids any third party software dependency. 
+
```

Comment:
By looking at this ADR it seems for me that Flux might fit a lot for a tooling purpose, so might be valid we either consider another option: using Flux in combination with the CLI prototype and have an Operator that specifically manages migrations. Following the main pros vs cons that I saw:

pros:
- harnessing the benefits of Flux and its capabilities
- as well as having a dedicated Operator to handle seamless migration management.

cons:
- we will probably NOT able to do operations prior, during or after installing the solutions. This could result in scenarios where certain tasks cannot be fully automated and would require manual intervention, potentially through support. In such cases, we may find ourselves resorting to using the CLI to address these limitations.
-